addons/udes_sale_stock/models/sale_order.py

```diff
@@ -40,6 +44,88 @@ def _set_priority(self):
                 'priority': order.priority
             })
 
+    @api.model
+    def get_available_stock_locations(self):
+        """ Method returns stock locations that are considered (along with
+         their children) stock available for fulfilling orders. Should be
+        overridden where necessary """
+        return self.env.ref('stock.stock_location_stock')
+
+    @api.model
+    def get_available_quantity(self, product_id, locations):
+        """ Get available quantity of product_id within locations """
+        Stock = self.env['stock.quant']
+        domain = [('product_id', '=', product_id.id),
+                  ('location_id', 'child_of', locations.ids)]
+        quants = Stock.search(domain)
+        available_quantity = sum(quants.mapped('quantity')) - \
+                             sum(quants.mapped('reserved_quantity'))
+        return available_quantity
+
+    @api.model
+    def cancel_orders_without_availability(self):
+        """ From the current list of unconfirmed SO lines, cancel lines that
+        cannot be fulfilled with current stock holding """
+        Order_line = self.env['sale.order.line']
+
+        # Get order lines
+        order_lines = Order_line.search([('state', 'in', ['sale', 'draft'])])
+
+        # Get unreserved stock for each product in locations
+        location_ids = self.get_available_stock_locations()
+        stock = {}
+        for pid in order_lines.mapped('product_id'):
+                stock[pid] = stock.get(pid, 0) + \
+                             self.get_available_quantity(pid, location_ids)
+
```

Comment:
A [defaultdict](https://docs.python.org/2/library/collections.html#defaultdict-examples) could be used for `stock` in this function

Reply:
Agree.

Reply:
never used defaultdict before, good shout 👍